src/libteam/0004-libteam-Add-lacp-fallback-support-for-single-member-.patch

```diff
@@ -0,0 +1,151 @@
+@@ -138,6 +138,8 @@ struct lacp {
+ #define		LACP_CFG_DFLT_SYS_PRIO 0xffff
+ 		bool fast_rate;
+ #define		LACP_CFG_DFLT_FAST_RATE false
++		bool fallback;
++#define		LACP_CFG_DFLT_FALLBACK false
+ 		int min_ports;
+ #define		LACP_CFG_DFLT_MIN_PORTS 1
+ 		enum lacp_agg_select_policy agg_select_policy;
+@@ -272,6 +274,11 @@ static int lacp_load_config(struct teamd_context *ctx, struct lacp *lacp)
+ 		lacp->cfg.fast_rate = LACP_CFG_DFLT_FAST_RATE;
+ 	teamd_log_dbg("Using fast_rate \"%d\".", lacp->cfg.fast_rate);
+ 
++	err = teamd_config_bool_get(ctx, &lacp->cfg.fallback, "$.runner.fallback");
++	if (err)
++		lacp->cfg.fallback = LACP_CFG_DFLT_FALLBACK;
++	teamd_log_dbg("Using fallback \"%d\".", lacp->cfg.fallback);
++
+ 	err = teamd_config_int_get(ctx, &tmp, "$.runner.min_ports");
+ 	if (err) {
+ 		lacp->cfg.min_ports = LACP_CFG_DFLT_MIN_PORTS;
+@@ -308,9 +315,24 @@ static bool lacp_port_loopback_free(struct lacp_port *lacp_port)
+ 	return true;
+ }
+ 
++/*
++ * is_lacp_fallback_eligible - is lacp_port eligible to go into lacp fallback mode
++ *
++ * Return true if it is, false otherwise
++ */
++static bool is_lacp_fallback_eligible(struct lacp_port *lacp_port)
++{
++	teamd_log_dbg("%s fallback eligible state \"%d \" cfg \"%d\".",
++				lacp_port->tdport->ifname, lacp_port->state,
++				lacp_port->lacp->cfg.fallback);
++	return lacp_port->state == PORT_STATE_DEFAULTED && 
++		lacp_port->lacp->cfg.fallback;
++}
++
+ static bool lacp_port_selectable_state(struct lacp_port *lacp_port)
+ {
+-	if (lacp_port->state == PORT_STATE_CURRENT)
++	if (lacp_port->state == PORT_STATE_CURRENT ||
++	    is_lacp_fallback_eligible(lacp_port))
+ 		return true;
+ 	return false;
+ }
+@@ -318,7 +340,8 @@ static bool lacp_port_selectable_state(struct lacp_port *lacp_port)
+ static bool lacp_port_unselectable_state(struct lacp_port *lacp_port)
+ {
+ 	if (lacp_port->state == PORT_STATE_CURRENT ||
+-	    lacp_port->state == PORT_STATE_EXPIRED)
++	    lacp_port->state == PORT_STATE_EXPIRED ||
++	    is_lacp_fallback_eligible(lacp_port))
+ 		return false;
+ 	return true;
+ }
+@@ -1002,6 +1025,12 @@ static int lacp_port_link_update(struct lacp_port *lacp_port)
+ 	uint32_t speed = team_get_port_speed(team_port);
+ 	uint8_t duplex = team_get_port_duplex(team_port);
+ 	int err;
++	bool admin_state;
++
++	admin_state = team_get_ifinfo_admin_state(lacp_port->ctx->ifinfo);
++	teamd_log_dbg("%s admin_state  \"%d\" link_up \"%d\".",
++				lacp_port->tdport->ifname,
++				admin_state, linkup);
+ 
+ 	if (linkup != lacp_port->__link_last.up ||
+ 	    duplex != lacp_port->__link_last.duplex) {
+@@ -1011,8 +1040,9 @@ static int lacp_port_link_update(struct lacp_port *lacp_port)
+ 		 * will provide speed == 0 and duplex == 0. If that is the
+ 		 * case now, do not set disabled state and allow such devices
+ 		 * to work properly.
++		 * Only enable port if the it is admin_up and link_up.
+ 		 */
+-		if (linkup && (!duplex == !speed))
++		if (linkup && (!duplex == !speed) && admin_state)
```

Comment:
Thanks Marian @marian-pritsak for the code review.

Yes, I removed the lacp fallback timer, which is the transition time from defaulted state to fallback state.  Since we reached an agreement to not modify existing LACP receive state machine, I removed the fallback state and the fallback timer is no longer needed.

We need to check the admin state because if we create a Lag PortChannel01 with member ports Ethernet1. The PortChannel will be by default admin Shut, but the member port will be in admin UP state. If we don't check the admin status of the LAG, the member will be moved into defaulted state since it's link up, and then be moved to fallback state if lacp fallback is enabled.

Please let me know if you still have questions.

Thanks,
Haiyang

Reply:
@marian-pritsak Hi Marian, Please let me know if you still have concerns regarding this admin check. If so, we can setup a skype call to answer your questions.

Thanks,
Haiyang

Reply:
@hzheng5 , this is actually separate issue. whether or not member port admin status depends on the lag admin status is a separate issue. i tested on arista switch, when the lag is down, the member port is errdisabled. is this issue related to lacp fallback mechanism? can you explain?